### PR TITLE
feat: burn into

### PR DESCRIPTION
**Pull request by Ad96el**
## fixes[#3660](https://github.com/KILTprotocol/ticket/issues/3660)
Includes the `burn_from` tx. 

For the burn operation, the lock on the account is removed. According to the current [thaw implementation](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/assets/src/lib.rs#L1088), this should work regardless of whether the account is active or frozen.

### CODE REVIEW

File: pallets/pallet-bonded-coins/src/lib.rs [hunk 1]
```diff
@@ -253,8 +253,84 @@ pub mod pallet {
...
+
+			// just remove any locks, if existing.
+			T::Fungibles::thaw(&admin, &beneficiary, target_currency_id).map_err(|freeze_error| freeze_error.into())?;
+
+			T::Fungibles::burn_from(
+				target_currency_id.clone(),
+				&beneficiary,
+				amount_to_burn,
+				TokenPrecision::Exact,
+				Fortitude::Polite,
```
Comment (rflechtner) [inline pallets/pallet-bonded-coins/src/lib.rs:329]: I always went for force - what do you think is best here? (ignoring the fact that the assets pallet simply ignores this flag)

File: pallets/pallet-bonded-coins/src/lib.rs [hunk 2]
```diff
@@ -253,8 +253,84 @@ pub mod pallet {
...
+		#[pallet::call_index(2)]
+		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
+		pub fn burn_into(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			currency_idx: u32,
+			amount_to_burn: FungiblesBalanceOf<T>,
+			min_return: CollateralCurrencyBalanceOf<T>,
+			beneficiary: AccountIdLookupOf<T>,
+		) -> DispatchResult {
```
Comment (rflechtner) [inline pallets/pallet-bonded-coins/src/lib.rs:268]: the currencies_count parameter is still missing

File: pallets/pallet-bonded-coins/src/lib.rs [hunk 3]
```diff
@@ -253,8 +253,84 @@ pub mod pallet {
...
+				pool_details.denomination,
+				currency_idx,
+			)?;
+
+			let normalized_amount_to_burn =
+				convert_to_fixed::<T>(amount_to_burn.saturated_into::<u128>(), pool_details.denomination)?;
+
+			let low = high
+				.checked_sub(normalized_amount_to_burn)
+				.ok_or(ArithmeticError::Overflow)?;
```
Comment (rflechtner) [inline pallets/pallet-bonded-coins/src/lib.rs:295]: looks like an underflow to me

**Review by rflechtner — COMMENTED** [review]
Beautiful! Just needs the currencies_count parameter added if I'm not mistaken

File: pallets/pallet-bonded-coins/src/lib.rs [hunk 4]
```diff
@@ -253,8 +253,84 @@ pub mod pallet {
...
+			pool_id: T::PoolId,
+			currency_idx: u32,
+			amount_to_burn: FungiblesBalanceOf<T>,
+			min_return: CollateralCurrencyBalanceOf<T>,
+			beneficiary: AccountIdLookupOf<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let beneficiary = T::Lookup::lookup(beneficiary)?;
+
+			let pool_details = Pools::<T>::get(pool_id.clone()).ok_or(Error::<T>::PoolUnknown)?;
```
Comment (ntn-x2) [inline pallets/pallet-bonded-coins/src/lib.rs:272]:
```suggestion
			let pool_details = Pools::<T>::get(&pool_id).ok_or(Error::<T>::PoolUnknown)?;
```

File: pallets/pallet-bonded-coins/src/lib.rs [hunk 5]
```diff
@@ -253,8 +253,84 @@ pub mod pallet {
...
+					log::error!(
+						target: LOG_TARGET,
+						"Admin not found for currency id: {:?}",
+						target_currency_id
+					);
+					Error::<T>::Internal
+				})?;
+
+			// just remove any locks, if existing.
+			T::Fungibles::thaw(&admin, &beneficiary, target_currency_id).map_err(|freeze_error| freeze_error.into())?;
```
Comment (ntn-x2) [inline pallets/pallet-bonded-coins/src/lib.rs:322]: Does this remove all locks, including ones that might have been set on the currency elsewhere? I don't see any lock ID, so I assume that is the case.

File: pallets/pallet-bonded-coins/src/lib.rs (hunk 1 above)
Comment (ntn-x2) [reply pallets/pallet-bonded-coins/src/lib.rs:329]: Yes, I also agree force is what we want here anyway.

File: pallets/pallet-bonded-coins/src/lib.rs [hunk 6]
```diff
@@ -253,8 +253,84 @@ pub mod pallet {
...
+				T::CollateralAssetId::get(),
+				&pool_id.into(),
+				&beneficiary,
+				collateral_return,
+				Preservation::Expendable,
+			)?;
+
+			// we act on behalf of the admin.
+			let admin = T::Fungibles::admin(target_currency_id.clone())
+				// Should never fail. Either the admin has been updated or it is the pool id.
```
Comment (ntn-x2) [inline pallets/pallet-bonded-coins/src/lib.rs:311]: What does this comment mean? Especially the second part about the pool ID?

File: pallets/pallet-bonded-coins/src/lib.rs [hunk 7]
```diff
@@ -253,8 +253,84 @@ pub mod pallet {
...
+			amount_to_burn: FungiblesBalanceOf<T>,
+			min_return: CollateralCurrencyBalanceOf<T>,
+			beneficiary: AccountIdLookupOf<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let beneficiary = T::Lookup::lookup(beneficiary)?;
+
+			let pool_details = Pools::<T>::get(pool_id.clone()).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(pool_details.is_minting_authorized(&who), Error::<T>::Locked);
```
[the remark this answers is not in the document]
Comment (rflechtner) [reply pallets/pallet-bonded-coins/src/lib.rs:274]: I thought the same, and I suggest we rename the method to `can_mint`. 
Minting is privileged if the pool is locked, so we could either throw NoPermission or Locked, but I think the latter is a bit more useful for 99% of users.

File: pallets/pallet-bonded-coins/src/lib.rs (hunk 5 above)
Comment (rflechtner) [reply pallets/pallet-bonded-coins/src/lib.rs:322]: The assets pallet has an account freeze, which is what we use here. I disallows any withdrawal of funds from an account.

File: pallets/pallet-bonded-coins/src/lib.rs (hunk 1 above)
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/lib.rs]: [4c8abbf](https://github.com/KILTprotocol/kilt-node/pull/771/commits/4c8abbf98137ee4fea00020b24668bb0e3383912)

File: pallets/pallet-bonded-coins/src/lib.rs (hunk 4 above)
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/lib.rs:289]: [3c78bb9](https://github.com/KILTprotocol/kilt-node/pull/771/commits/3c78bb95e30febe381dd7c138dd8a05f9fa90920)

File: pallets/pallet-bonded-coins/src/lib.rs (hunk 2 above)
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/lib.rs]: True. updated in: https://github.com/KILTprotocol/kilt-node/pull/771/files#diff-114bec8b58b7c7a182ca4abd208a4459ebd849a5e8852dddca197c043fcb1ae4R284

File: pallets/pallet-bonded-coins/src/lib.rs [hunk 8]
```diff
@@ -253,8 +253,84 @@ pub mod pallet {
...
+
+			let bonded_currencies = pool_details.bonded_currencies;
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/lib.rs:293]: Thanks for the hint!

File: pallets/pallet-bonded-coins/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/lib.rs]: I updated the error to `NoPermission`.  Should I also update the function? I think `is_minting_authorized` is fine.

File: pallets/pallet-bonded-coins/src/lib.rs (hunk 6 above)
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/lib.rs:333]: Either the pool_id (also the pool account) is the manager, or it was updated in the reset_team transaction. What specifically is unclear for you? I can update the comment to clarify.

Comment (ntn-x2) [reply pallets/pallet-bonded-coins/src/lib.rs:333]: I mean the whole sentence `// Should never fail. Either the admin has been updated or it is the pool id.` is a bit unclear. The second part means that the pool ID is the manager? Maybe you can add that?

**Review by rflechtner — APPROVED** [review]
LGTM!

File: pallets/pallet-bonded-coins/src/lib.rs (hunk 6 above)
Comment (rflechtner) [reply pallets/pallet-bonded-coins/src/lib.rs]: that's gone now I think

**Pull request merged**